src/bridge/ex_data_util/ex_data_util.cc

```diff
@@ -0,0 +1,151 @@
+
+namespace kmsengine {
+namespace bridge {
+namespace {
+
+// Represents an uninitialized OpenSSL external index. Value is -1 since
+// OpenSSL's `CRYPTO_get_ex_new_index` function for requesting external indicies
+// returns -1 on failure.
+static constexpr int kUninitializedIndex = -1;
+
+// External index assigned by OpenSSL on a `RSA` struct. If uninitialized, it
+// has value `kUninitializedIndex`. Used in `AttachRsaKeyToOpenSslRsa` and
+// `GetRsaKeyFromOpenSslRsa`.
+static int rsa_index = kUninitializedIndex;
+
+// External index assigned by OpenSSL on a `ENGINE` struct. If uninitialized, it
+// has value `kUninitializedIndex`. Used in `AttachEngineDataToOpenSslEngine`
+// and `GetEngineDataFromOpenSslEngine`.
+static int engine_index = kUninitializedIndex;
+
+// Requests an external index from OpenSSL for the index type `index_type`.
+// Valid index types are the `CRYPTO_EX_INDEX_*` constants found in OpenSSL's
+// crypto.h header.
+//
+// Returns the external index on success or an error `Status`.
+StatusOr<int> GetIndex(int index_type) {
+  // We ignore the argl and argp parameters (the second and third parameters)
+  // since they're used in the callback functions (the fourth, fifth, and
+  // sixth parameters), which we're just setting to `nullptr`.
+  int index = CRYPTO_get_ex_new_index(index_type, 0, nullptr, nullptr,
+                                      nullptr, nullptr);
+  if (index == kUninitializedIndex) {
+    return Status(StatusCode::kInternal, "No CRYPTO_EX_DATA index available");
+  }
+  return index;
+}
+
+// Returns `rsa_index` if it is initialized, or an error `Status`.
+inline StatusOr<int> GetRsaIndex() {
+  if (rsa_index == kUninitializedIndex) {
+    return Status(StatusCode::kFailedPrecondition, "rsa_index uninitialized");
+  }
+  return rsa_index;
+}
+
+// Returns `engine_index` if it is initialized, or an error `Status`.
+inline StatusOr<int> GetEngineIndex() {
+  if (engine_index == kUninitializedIndex) {
+    return Status(StatusCode::kFailedPrecondition,
+                  "engine_index uninitialized");
+  }
+  return engine_index;
+}
+
+}  // namespace
+
+Status InitExternalIndicies() {
+  KMSENGINE_ASSIGN_OR_RETURN(rsa_index, GetIndex(CRYPTO_EX_INDEX_RSA));
+  KMSENGINE_ASSIGN_OR_RETURN(engine_index, GetIndex(CRYPTO_EX_INDEX_ENGINE));
+  return Status();
+}
+
+void FreeExternalIndicies() {
+  CRYPTO_free_ex_index(CRYPTO_EX_INDEX_RSA, rsa_index);
+  CRYPTO_free_ex_index(CRYPTO_EX_INDEX_ENGINE, engine_index);
+  rsa_index = kUninitializedIndex;
+  engine_index = kUninitializedIndex;
+}
+
+Status AttachRsaKeyToOpenSslRsa(backing::RsaKey *rsa_key, RSA *rsa) {
+  if (rsa == nullptr) {
+    return Status(StatusCode::kInvalidArgument, "RSA cannot be null");
+  }
+
+  KMSENGINE_ASSIGN_OR_RETURN(auto index, GetRsaIndex());
+  if (!RSA_set_ex_data(rsa, index, static_cast<void *>(rsa_key))) {
+    return Status(StatusCode::kInternal, "RSA_set_ex_data failed");
+  }
+  return Status();
```

Comment:
I think it would be better not to rely on the default constructor of `Status`. Is there an `OkStatus()` method of some sort that you can use?